libmesabox/src/lsb/mount/mod.rs

```diff
@@ -0,0 +1,803 @@
+        option_map.insert(OsString::from("ro"), libc::MS_RDONLY);
+        option_map.insert(OsString::from("rw"), !libc::MS_RDONLY);
+        option_map.insert(OsString::from("suid"), !libc::MS_NOSUID);
+        option_map.insert(OsString::from("sync"), libc::MS_SYNCHRONOUS);
+        option_map
+    };
+}
+
+struct Flag {
+    flag: MsFlags,
+}
+
+impl Default for Flag {
+    fn default() -> Self {
+        Self {
+            flag: MsFlags::MS_SILENT,
+        }
+    }
+}
+
+impl Flag {
+    fn from(options: &mut Vec<OsString>) -> MountResult<MsFlags> {
+        //let mut flag = Flag::default();
+        let mut flag = Flag::default().flag.bits();
+        if options.contains(&OsString::from("default")) {
+            options.extend_from_slice(&[
+                OsString::from("rw"),
+                OsString::from("suid"),
+                OsString::from("dev"),
+                OsString::from("exec"),
+                OsString::from("auto"),
+                OsString::from("nouser"),
+                OsString::from("async"),
+            ]);
+        }
+        for opt in options {
+            let f = *OPTION_MAP.get(opt).ok_or(MountError::UnsupportedOption)?;
+            if (f as c_long) < 0 {
+                flag &= f;
+            } else {
+                flag |= f;
+            }
+        }
+        Ok(MsFlags::from_bits(flag).ok_or(MountError::UnsupportedOption)?)
+    }
+}
+
+struct MntEnt {
+    mnt_fsname: OsString,
+    mnt_dir: OsString,
+    mnt_type: OsString,
+    mnt_opts: OsString,
+}
+
+/// This is used to read Filesystem Description File
+/// e.g. /etc/fstab and /etc/mtab
+struct FSDescFile {
+    entries: Vec<MntEnt>,
+}
+
+impl FSDescFile {
+    fn new(path: &str) -> MountResult<Self> {
+        // all of these files should exist and can be read, but just in case
+        let file = fs::File::open(path).or(Err(MountError::OpenFileError(String::from(path))))?;
+        let mut entries = Vec::new();
+        for line in BufReader::new(file).lines() {
+            let line = line.or(Err(MountError::FSDescFileFormatError(String::from(path))))?;
+            match line.chars().next() {
+                None | Some('#') => {
+                    continue;
+                }
+                Some(_) => {}
+            }
+            let a: Vec<&str> = line.split_whitespace().collect();
+            // There should be 6 columns in FileSystem Description Files
+            if a.len() != 6 {
+                return Err(MountError::FSDescFileFormatError(String::from(path)));
+            }
+            // We only need the first 4 columns
+            let mnt = MntEnt {
+                mnt_fsname: OsString::from(a[0]),
+                mnt_dir: OsString::from(a[1]),
+                mnt_type: OsString::from(a[2]),
+                mnt_opts: OsString::from(a[3]),
+            };
+            entries.push(mnt)
+        }
+        Ok(Self { entries: entries })
+    }
+
+    fn get_output(&self, fs_type: &OsString) -> String {
```

Comment:
Again, `String` seems wrong because the filesystem directory doesn't need to be valid UTF-8.

Reply:
The returned string is used for output, and it is generated by `to_string_lossy`. I guess this is the only way to get a printable `OsString`.